Lib/asyncio/tasks.py

```diff
@@ -496,7 +496,13 @@ async def wait_for(fut, timeout, *, loop=None):
             # after wait_for() returns.
             # See https://bugs.python.org/issue32751
             await _cancel_and_wait(fut, loop=loop)
-            raise exceptions.TimeoutError()
+            # In case task cancellation produced some result
+            # (either value or exception), we should pass through it
+            # See https://bugs.python.org/issue40607
+            try:
+                return fut.result()
+            except exceptions.CancelledError as exc:
+                raise exceptions.TimeoutError() from exc
```

Comment:
I'd put also an `else: raise TimeoutError`

Reply:
Sorry, but I think that `else` branch is unreachable: if there isn't any exception in try block, then `return` works normally and we leave this method. Please explain if I am wrong.

Reply:
Yes, but your new code is simply not equivalent to the previous code that raised TimeoutError unconditionally. Now I see that there might be a condition for clear return and I don't want that.

Reply:
The linked issue talks about the correct exception propagation. Basically, `wait_for` timeouts with an exception and we want that original exception to be propagated. That's fine and I'm +1 for that.

The code here is slightly different though and a bit hard to trace so it would be easier to read if it was replaced with:

```python
           try:
                fut.result()
            except exceptions.CancelledError as exc:
                raise exceptions.TimeoutError() from exc
            else:
                raise exceptions.TimeoutError()
```

Reply:
I think the issue here is that there is a race condition where the task can complete after the timeout has elapsed. So the question is whether the caller should get to learn that the task completed in this case.

Reply:
Yes, the current implementation discards the result, as the probability of actually having it is low. I'd like to keep it that way. If we want to challenge that current behavior that should be done in a separate issue, with a new PR and new discussion.